text/0042-partial-fill-rules.md

```diff
@@ -0,0 +1,296 @@
+     numerator and denominator. This equation must be validated by clearing denominators and checking an inequality of `u128`'s, to
+     avoid numerical issues.
+     
+     ```
+     fractional_output.value * fill_fraction_denominator >= fill_fraction_numerator * partial_fill_output.value
+     ```
+     
+     If this inequality does not hold, then the transaction must be rejected for reasons of not respecting the fill fraction.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This proposal has a few drawbacks.
+
+* The amounts of partial fill transactions are revealed to the enclave. Currently, MobileCoin transactions are validated using RingCT and do not reveal amounts to the enclave. It would be better if there were a way to validate partial fills without revealing this to the enclave, but for the use-cases envisioned, this data already has to be revealed anyways, so the use-cases aren't negatively impacted. We could imagine that an exchange service is based on SGX, and so the amounts of these quotes are only revealed inside of SGX, and are matched inside of SGX. Even in this case, it is still the case that if you penetrate SGX you will see the amounts of the SCI quotes.
+* Partial-fill SCI quotes cannot be re-used by a second counterparty later if the first counter-party only matched some of the volume.
+In a traditional exchange, limit orders that are partially-filled remain on the books until being totally filled, so that potentially dozens of different parties can each match a piece of the liquidity provided by the originator. In this proposal, that would not work, because as soon as the first counterparty matches the order, the key image underlying the SCI is burned. One could imagine that there might be a way that the fractional change output which goes back to the originator could be automatically signed as an SCI somehow, so that matching can continue without the originator being in the loop. This proposal instead envisions that if the originator wants to continue matching they simply sign a new SCI themselves and re-broadcast.
+* The same input can be used to "back" multiple SCI quotes at once, in which case if any one of them is matched, the others instantly become impossible to match. This is somewhat different from how limit orders work in an exchange -- each limit order must be backed by distinct liquidity. This reflects the idea that SCI's are more in analogy with a signed quote than with a limit order as such. This is potentially confusing to e.g. a trader who looks at a list of available SCI quotes and infers that there is a certain amount of liquidity depth, reasoning as they would in a traditional exchange. But if the same inputs are used to sign multiple quotes e.g. for different currencies, it may be impossible that all of these quotes could actually match, and the "real" liquidity may be considerably less. This can be mitigated by making sure that the software they are using communicates this idea clearly.
```

Comment:
> What ensures this, is it the key image specifically tied to the txout.

Yes -- each TxOut has a unique KeyImage. This is revealed when an MLSAG spending that TxOut is signed, and there is no way to make a valid MLSAG signature that reveals something other than that one true key image. This is the mechanism that prevents double spends in MobileCoin.

An SCI, containing one MLSAG, also has a key image. Nothing stops you from signing more than one MLSAG or SCI over the same TxOut, but they all become mutually exclusive because transaction validation prevents a key image from appearing twice in the ledger.

> If someone doesn't have a full amount would they first need to make a tx that combines their chunks into one?

Not necessarily I think? If they have 3 inputs, they could sign 3 different SCIs all with the same implied price? They could consolidate the Tx's first

> Is there a crypto term for combining, smelt..??

I am not sure if there's a standard term, I've used the term "consolidate", we also have a concept of "wallet fragmentation" errors and a "defragmentation Tx" to fix it. There's a common term "dust" that I know of.

:laughing: 
Whoever smelt it dealt it

Reply:
> Whoever smelt it dealt it

:)